From 22bc1233af3a2c8d39731fe2ab5bb0b212a1fb32 Mon Sep 17 00:00:00 2001
From: Jonas Jelonek <jelonek.jonas@gmail.com>
Date: Mon, 16 Jan 2023 17:03:15 +0100
Subject: [PATCH 2/7] mac80211: add tx-power annotation in control path

Adds members to ieee80211_tx_info and ieee80211_sta_rates structures
to allow tx-power annotation per packet/per mrr stage. The added members
are always tx-power indices referring to the tx-power set described by
ieee80211_hw->txpower_ranges.

The annotation in ieee80211_tx_info is for probing and compatibility
reasons only, e.g. drivers that only support RC/TPC per packet and do
not yet use ieee80211_sta_rates.

Signed-off-by: Thomas Huehn <thomas.huehn@hs-nordhausen.de>
Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
---
 include/net/mac80211.h | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1106,6 +1106,10 @@ ieee80211_rate_get_vht_nss(const struct
  * @control.use_cts_prot: use RTS/CTS
  * @control.short_preamble: use short preamble (CCK only)
  * @control.skip_table: skip externally configured rate table
+ * @control.txpower_idx: Tx-power level index for whole packet, referring
+ * 	to an idx described by ieee80211_hw->txpower_ranges. A negative
+ * 	idx means 'invalid', 'unset', or 'default'. Behavior in this case
+ * 	is driver-specific.
  * @control.jiffies: timestamp for expiry on powersave clients
  * @control.vif: virtual interface (may be NULL)
  * @control.hw_key: key to encrypt with (may be NULL)
@@ -1154,7 +1158,7 @@ struct ieee80211_tx_info {
 					u8 use_cts_prot:1;
 					u8 short_preamble:1;
 					u8 skip_table:1;
-					/* 2 bytes free */
+					s16 txpower_idx;
 				};
 				/* only needed before rate control */
 				unsigned long jiffies;
@@ -2149,6 +2153,10 @@ enum ieee80211_sta_rx_bandwidth {
  * @rcu_head: RCU head used for freeing the table on update
  * @rate: transmit rates/flags to be used by default.
  *	Overriding entries per-packet is possible by using cb tx control.
+ * @rate.txpower_idx: An idx pointing to a tx-power level described by
+ * 	ieee80211_hw->txpower_ranges that should be used for the mrr stage.
+ * 	A negative value means 'invalid', 'unset' or 'default' power level,
+ * 	actual behavior is driver-specific.
  */
 struct ieee80211_sta_rates {
 	struct rcu_head rcu_head;
@@ -2158,6 +2166,7 @@ struct ieee80211_sta_rates {
 		u8 count_cts;
 		u8 count_rts;
 		u16 flags;
+		s16 txpower_idx;
 	} rate[IEEE80211_TX_RATE_TABLE_SIZE];
 };
 
